### ISSUE TITLE
[Feature] MLX Graph Runner — Design & Implementation Plan

### ISSUE THREAD

**Issue by shvmjndl** [sgl-project/sglang#19146]
### Checklist

- [x] If this is not a feature request but a general question, please start a discussion at https://github.com/sgl-project/sglang/discussions. Otherwise, it will be closed.
- [x] Please use English. Otherwise, it will be closed.

### Motivation

This issue tracks the design and implementation of `MLXGraphRunner` — the graph runner component required for the MLX backend on Apple Silicon, as specified in #17846 (architecture) and #19137 (roadmap).

#### Why a Graph Runner Is Needed

In SGLang, every device backend requires a **graph runner** — the component that executes the model's forward pass on the fast path during inference. The `ModelRunner` dispatches to the appropriate graph runner based on device type (`model_runner.py:2149-2156`):

```python
graph_runners = defaultdict(
    lambda: CudaGraphRunner,
    {
        "cpu": CPUGraphRunner,
        "npu": NPUGraphRunner,
    },
)
self.graph_runner = graph_runners[self.device](self)
```

Without an MLX graph runner registered here, the MLX backend cannot use the optimized forward path. `mlx_graph_runner.py` is listed as **REQUIRED for MVP** in #17846.

#### Why MLX Needs Its Own Runner

The existing runners rely on APIs that don't exist on Apple Silicon:
- **CudaGraphRunner**: `torch.cuda.CUDAGraph`, CUDA streams, GPU memory pools, per-batch-size graph capture/replay (~700 lines of CUDA-specific logic)
- **CPUGraphRunner**: `torch.compile` with CPU-specific fake ops registration

MLX's architecture is fundamentally different:
- **Unified memory** — no CPU↔GPU transfers, eliminates the need for CUDA-style capture/replay
- **Lazy evaluation** — computation deferred until `mx.eval()` is called
- **JIT compilation** — `mx.compile()` provides kernel fusion and Metal shader optimization, replacing CUDA graphs' role of batching kernel launches

This means the MLX graph runner is **significantly simpler** — a thin wrapper using `mx.compile()` rather than the complex capture/replay machinery.

---

#### File Location

```
python/sglang/srt/hardware_backend/mlx/mlx_graph_runner.py
```

Per the project structure defined in #17846.

#### Required Interface

There is no formal base class — graph runners follow a duck-typed contract. The three methods expected by `ModelRunner`:

| Method | Signature | Called From |
|--------|-----------|-------------|
| `__init__` | `(model_runner: ModelRunner)` | `init_device_graphs()` — `model_runner.py:2156` |
| `can_run` | `(forward_batch: ForwardBatch) -> bool` | `_forward_raw()` — `model_runner.py:2456-2459` |
| `replay` | `(forward_batch, skip_attn_backend_init=False, pp_proxy_tensors=None) -> LogitsProcessorOutput \| PPProxyTensors` | `_forward_raw()` — `model_runner.py:2463-2467` |

Additionally, `self.bs` is accessed externally via `getattr(self.graph_runner, "bs", None)` at `model_runner.py:2435` for expert caching.

#### Architecture

```
MLXGraphRunner (standalone class — does NOT inherit CudaGraphRunner)
│
├── __init__(model_runner)
│   ├── Store model_runner reference
│   ├── Validate unsupported features (LoRA, speculative, PP, DP, TP, etc.)
│   ├── Set capture_hidden_mode
│   └── self.compiled_fn = mx.compile(model.forward)
│
├── can_run(forward_batch) -> bool
│   └── Check capture_hidden_mode compatibility
│
└── replay(forward_batch, ...) -> LogitsProcessorOutput | PPProxyTensors
    ├── Set self.bs for external access
    ├── Init attention backend metadata (if not skipped)
    ├── Call self.compiled_fn(input_ids, positions, forward_batch)
    └── Return output
```

#### Key Design Decisions

**1. Standalone class, not subclassing CudaGraphRunner**

The NPU runner subclasses `CudaGraphRunner` because NPU graphs work like CUDA graphs (capture/replay). MLX has no such concept — inheriting would bring ~700 lines of unused CUDA-specific code. `CPUGraphRunner` is also standalone for the same reason.

**2. `mx.compile()` wraps `model.forward` once at init**

Per #17846:
```python
self.compiled_fn = mx.compile(self.model.forward)
```

Unlike CUDA graphs which require capturing a separate graph per batch size (`self.graphs[bs]`), `mx.compile()` handles varying shapes with a single compiled function. No per-batch-size capture loop needed.

**3. No pre-allocated buffers or padding**

CudaGraphRunner pre-allocates fixed-size input/output buffers (`DecodeInputBuffers`) and pads batches to captured sizes. MLX's unified memory and dynamic compilation make this unnecessary.

**4. `can_run()` has no batch-size restriction**

Since there's no per-batch-size graph dict to look up, `can_run()` only checks `capture_hidden_mode` compatibility — same logic as CPUGraphRunner (`cpu_graph_runner.py:446-463`).

**5. MVP limitations**

The initial implementation will assert these are not enabled (matching CPUGraphRunner's restrictions):
- LoRA
- Speculative decoding (EAGLE, ngram)
- Pipeline parallelism (PP > 1)
- Data parallelism (DP > 1)
- Tensor parallelism (TP > 1)
- Two-batch overlap
- Encoder-decoder models

These can be relaxed in later phases.

#### Registration Changes Required

| File | Change |
|------|--------|
| `device_config.py:14` | Add `"mlx"` to valid device list |
| `model_runner.py` (imports) | Add `from sglang.srt.hardware_backend.mlx.mlx_graph_runner import MLXGraphRunner` |
| `model_runner.py:2149-2155` | Add `"mlx": MLXGraphRunner` to `graph_runners` dict |
| `model_runner.py:2131-2135` | Handle MLX in the init gate logic (currently blocks non-CPU devices when `disable_cuda_graph=True`) |
| `model_runner.py:2451-2454` | Add `"mlx"` to `mode_check` — reuse `is_cpu_graph` (DECODE only) or add `is_mlx_graph` to `ForwardMode` |

#### How CudaGraphRunner vs CPUGraphRunner vs MLXGraphRunner Compare

| Aspect | CudaGraphRunner | CPUGraphRunner | MLXGraphRunner |
|--------|----------------|----------------|----------------|
| Optimization | CUDA graph capture/replay | `torch.compile` | `mx.compile()` JIT |
| Memory model | Discrete GPU, explicit transfers | CPU RAM | Unified memory (no transfers) |
| Per-batch-size graphs | Yes (`self.graphs[bs]`) | Yes (`self.graphs[bs]`) | **No** — single compiled fn |
| Warm-up captures | Iterates all batch sizes | Iterates all batch sizes | **None** — single compile call |
| Buffer pre-allocation | Yes (fixed CUDA buffers) | Yes (fixed CPU buffers) | **No** |
| Inherits CudaGraphRunner | N/A (is the base) | No (standalone) | No (standalone) |

---

### Dependencies

- `mlx_attention.py` — the graph runner calls `attn_backend.init_forward_metadata()`
- `mlx_model_loader.py` — model must be loaded before the graph runner initializes
- `device_config.py` — `"mlx"` must be recognized as a valid device

### Testing Plan

1. **Unit test**: Mock `ModelRunner`, verify `can_run()` / `replay()` interface contract
2. **Integration test**: End-to-end inference with `mlx-community/TinyLlama-1.1B-Chat-v1.0-4bit`
3. **Correctness**: Compare output logits against reference MLX-LM implementation

### Related

- Architecture: #17846
- Roadmap: #19137
- Reference implementations: `cpu_graph_runner.py`, `npu_graph_runner.py`
- PR target: `sglang-mlx/sglang` branch `feature/mlx-backend`
- MLX compile docs: https://ml-explore.github.io/mlx/build/html/usage/compile.html
